physutils/io.py

```diff
@@ -0,0 +1,235 @@
+        phys = physio.Physio(**inp)
+    # if we got a numpy array, load that into a Physio object
+    elif isinstance(data, np.ndarray):
+        logger.debug("Instantiating Physio object from numpy array")
+        if history is None:
+            logger.warning(
+                "Loading data from a numpy array without providing a"
+                "history will render reproducibility functions "
+                "useless! Continuing anyways."
+            )
+        phys = physio.Physio(np.asarray(data, dtype=dtype), fs=fs, history=history)
+    # create a new Physio object out of a provided Physio object
+    elif isinstance(data, physio.Physio):
+        logger.debug(
+            "Instantiating a new Physio object from the provided Physio object"
+        )
+        phys = physio.new_physio_like(data, data.data, fs=fs, dtype=dtype)
+        phys._history += [physio._get_call()]
+    else:
+        raise TypeError("Cannot load data of type {}".format(type(data)))
+
+    # reset sampling rate, as requested
+    if fs is not None and fs != phys.fs:
+        if not np.isnan(phys.fs):
+            logger.warning(
+                "Provided sampling rate does not match loaded rate. "
+                "Resetting loaded sampling rate {} to provided {}".format(phys.fs, fs)
+            )
+        phys._fs = fs
+    # coerce datatype, if needed
+    if dtype is not None:
+        phys._data = np.asarray(phys[:], dtype=dtype)
+
+    return phys
+
+
+def save_physio(fname, data):
+    """
+    Saves `data` to `fname`
+
+    Parameters
+    ----------
+    fname : str
+        Path to output file; .phys will be appended if necessary
+    data : Physio_like
+        Data to be saved to file
+
+    Returns
+    -------
+    fname : str
+        Full filepath to saved output
+    """
+
+    from physutils.physio import check_physio
+
+    data = check_physio(data)
+    fname += ".phys" if not fname.endswith(".phys") else ""
+    with open(fname, "wb") as dest:
+        hist = data.history if data.history != [] else None
+        np.savez_compressed(
+            dest, data=data.data, fs=data.fs, history=hist, metadata=data._metadata
+        )
+    logger.info(f"Saved {data} in {fname}")
+
+    return fname
+
+
+def load_history(file, verbose=False):
+    """
+    Loads history from `file` and replays it, creating new Physio instance
+
+    Parameters
+    ----------
+    file : str
+        Path to input JSON file
+    verbose : bool, optional
+        Whether to print messages as history is being replayed. Default: False
+
+    Returns
+    -------
+    file : str
+        Full filepath to saved output
+    """
+
+    # import inside function for safety!
+    # we'll likely be replaying some functions from within this module...
+
+    # TODO: These will need to be imported in order to replay history from this module. Unless another way is found
+    # import peakdet
+    # import phys2denoise
```

Comment:
We'll need to be cautious in how/whether we implement this to avoid introducing unnecessary dependencies! I think it will be good to open another issue to address this.

Reply:
In [ff22aec](https://github.com/physiopy/physutils/pull/2/commits/ff22aec5a7cc9950e678c3792071fbb37b7fc4c1) a warning is added for not installed modules. Would it be useful to implement for example:
- Having a history.json that uses both peakdet and physutils
- Only having installed peakdet
- Perform only the peakdet operations

I'd guess that the other way around wouldn't be useful. Still, even this case wouldn't seem too useful either, but it could make sense to implement

Wdyt @m-miedema ?

Reply:
Sorry, I missed this comment until now @maestroque ! The way I'm understanding what you're proposing, I agree that it wouldn't be too useful to implement.